subsys/power/device.c

```diff
@@ -50,7 +50,8 @@ static const char *const core_devices[] = {
 	"sys_clock",
 };
 #else
-#error "Add SoC's core devices list for PM"
+#define MAX_PM_DEVICES CONFIG_MAX_PM_DEVICES
+__weak const char *const core_devices[] = {"",};
```

Comment:
I don't think this is a good API.  `core_devices` should not be a global symbol.  It could be changed to `z_core_devices` or something that's namespaced.

Also the relation between the array and its size is too fragile.  I think a case can be made to allow an external override via a weak symbol, but the number of devices should be detected by a terminal null pointer rather than a configuration-time constant.

Reply:
Thank you @pabigot , I once considered to do it like this:

extern const char *const *z_core_devices; (a null terminated array )
static const char* const *core_devices = z_core_devices;

then change all core_devices[] in this file to a null terminated array
in sys_pm_create_device_list() enumerate core_devices by "if(core_devices[i])"
ensure size of core_devices less than MAX_PM_DEVICES by assert.

what do you think about that?

Reply:
For the list of core devices specifically I think it can be slightly simpler and more flexible if you do something like the top commit of https://github.com/pabigot/zephyr/commits/proposed/28106.  That depends on a Kconfig `PM_MAX_DEVICES` similar to the one you have here (the name should start with a subsystem, could be `POWER_MAX_DEVICES`), and uses the `z_pm_core_devices` directly.  And simplifies how the defaults are set.

Something like that, at any rate.  You can incorporate that patch into your PR if you wish (it hasn't been tested at all).  The other ideas are fine.

This whole thing should go away soon so it's not too critical to get it perfect.

Reply:
Good idea @pabigot . I try to base my patch on your https://github.com/pabigot/zephyr/commits/proposed/28106 but failed, so I include your modifications in that patch into my patch, do you mind that?

Reply:
I don't mind, this looks much better.  Thanks.